README.md
<a href="https://www.buymeacoffee.com/ssilverman" title="Donate to this project using Buy Me A Coffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" alt="Buy Me A Coffee donate button"></a>

# Readme for TeensyDMX v4.3.0-snapshot

This is a full-featured library for receiving and transmitting DMX on Teensy 3,
Teensy LC, and Teensy 4. It follows the
[ANSI E1.11 DMX512-A specification](https://tsp.esta.org/tsp/documents/published_docs.php).

## Table of contents

1. [Features](#features)
   1. [Receiver timing limitations and RX line monitoring](#receiver-timing-limitations-and-rx-line-monitoring)
   2. [Transmitter timing limitations](#transmitter-timing-limitations)
2. [The TODO list](#the-todo-list)
3. [How to use](#how-to-use)
   1. [Examples](#examples)
   2. [Synchronous vs. asynchronous operation](#synchronous-vs-asynchronous-operation)
4. [DMX receive](#dmx-receive)
   1. [Code example](#code-example)
   2. [Retrieving 16-bit values](#retrieving-16-bit-values)
   3. [Error counts and disconnection](#error-counts-and-disconnection)
      1. [The truth about connection detection](#the-truth-about-connection-detection)
      2. [Keeping short packets](#keeping-short-packets)
   4. [Packet statistics](#packet-statistics)
   5. [Error statistics](#error-statistics)
   6. [Synchronous operation by using custom responders](#synchronous-operation-by-using-custom-responders)
      1. [Responding](#responding)
5. [DMX transmit](#dmx-transmit)
   1. [Code example](#code-example-1)
   2. [Packet size](#packet-size)
   3. [Transmission rate](#transmission-rate)
   4. [Synchronous operation by pausing and resuming](#synchronous-operation-by-pausing-and-resuming)
   5. [Choosing BREAK and MAB times](#choosing-break-and-mab-times)
      1. [Specific BREAK/MAB times](#specific-breakmab-times)
         1. [A note on BREAK timing](#a-note-on-break-timing)
         2. [A note on MAB timing](#a-note-on-mab-timing)
      2. [BREAK/MAB times using serial parameters](#breakmab-times-using-serial-parameters)
   6. [Inter-slot MARK time](#inter-slot-mark-time)
   7. [MBB time](#mbb-time)
   8. [Error handling in the API](#error-handling-in-the-api)
6. [Technical notes](#technical-notes)
   1. [Simultaneous transmit and receive](#simultaneous-transmit-and-receive)
   2. [Transmission rate](#transmission-rate)
   3. [Transmit/receive enable pins](#transmitreceive-enable-pins)
   4. [Thread safety](#thread-safety)
   5. [Dynamic memory allocation failures](#dynamic-memory-allocation-failures)
   6. [Hardware connection](#hardware-connection)
   7. [`Receiver` and driving the TX pin](#receiver-and-driving-the-tx-pin)
   8. [Potential PIT timer conflicts](#potential-pit-timer-conflicts)
7. [Code style](#code-style)
8. [References](#references)
9. [Acknowledgements](#acknowledgements)

## Features

Some notable features of this library:

1. Teensy's default serial buffer isn't used; the data goes directly to/from
   the DMX buffers from/to the UART ISRs. In other words, the library is
   asynchronous and runs independently; all you need to worry about is setting
   and getting channel data.
2. Simple API: After setup, there's only two read calls (`readPacket` and `get`)
   and two forms of one write call (`set` for single and multiple channels).
3. The library properly handles DMX packets containing less than 513 slots.
4. The transmitter refresh rate can be changed to something less than
   "maximum rate".
5. The transmitter can be paused and resumed to allow for packets that must be
   adjacent to other packets. In other words, the asynchronous transmitter can
   be used synchronously. For example, System Information Packets (SIP) require
   this. See Annex D5 of ANSI E1.11.
6. The transmitter timing parameters can be specified: BREAK, MAB, inter-slot
   MARK time, and MBB.
7. The receiver checks for timeouts according to the DMX specification. It
   knows of the concept of being disconnected from a DMX transmitter when
   timeouts or bad BREAKs are encountered in the data stream.
8. Packet and error statistics are available. These can be used to detect
   protocol problems, including timeouts, framing errors and bad BREAKs, short
   packets (those less than 1196us), and long packets (those that exceed
   513 bytes).
9. The receiver can be used synchronously through the use of the `Responder`
   API. Alternate start codes can not only be handled, for example, for Text
   Packets or System Information Packets (SIP), but responses can be sent back
   to the transmitter, for example for RDM.
10. Functions for handling 16-bit data.

### Receiver timing limitations and RX line monitoring

When the RX line is not being monitored, there are limitations in the handling
of received DMX frame timing. For BREAK and Mark after BREAK (MAB) times, only
the following cases are checked and not accepted as a valid DMX frame start:

1. BREAK duration &lt; ~44us.
2. BREAK duration + MAB duration &lt; ~96us.
3. BREAK &lt; ~88us and MAB &ge; ~44us.

The following case is accepted as a valid frame start, even though it isn't
compliant with the DMX specification:

1. BREAK duration &gt; ~52us and MAB duration &lt; ~44us.

For example, if a BREAK comes in having a duration of 53us and then a MAB comes
in having a duration of 43us, their sum is 96us, and so the packet will be
accepted. More generally, this will allow BREAKs that are shorter than the
minimum required 88us if the MAB is shorter than 44us.

This limitation does not exist if the RX line is monitored. To monitor the line,
connect it to a digital I/O-capable pin and call `setRXWatchPin` with the pin
number. The pin cannot be the same as the RX pin.

### Transmitter timing limitations

The transmitter uses a UART to control all the output. On the Teensy, the UART
runs independently. This means that any specified timings won't necessarily be
precise. They will often be accurate to within one or maybe two bit times. An
effort has been made, however, to make sure that transmitted timings are at
least as long as the requested timings.

For example, if a 100us MAB is requested, the actual MAB may be 102 or 103us. If
a 40us inter-slot time is requested, the actual time may be 44us. And so on.

Additionally, certain timings will have a minimum length that can't be
shortened, for example the MBB. This is simply due to code execution time
interacting with the interrupt and UART subsystems. For example, on a Teensy LC,
the minimum MBB is about 119us, even if the requested value is 0us.

## The TODO list

These are either in the works or ideas for subsequent versions:

1. Asynchronous responder data. Currently, the data is sent synchronously inside
   the UART ISR where responders process the packet.
2. Better MAB transmit timing, perhaps by somehow synchronizing with the baud
   rate clock.
3. Explore much more precise transmitter timings by not using the UART.

## How to use

The classes you'll need are in the `qindesign::teensydmx` namespace: `Receiver`
and `Sender`.

All class documentation can be found in `src/TeensyDMX.h`.

### Examples

`Receiver` examples:
* `BasicReceive`: A basic receive example
* `Flasher`: Change the flash speed of the board LED based on DMX input

`Sender` examples:
* `BasicSend`: A basic send example
* `Chaser`: Chases values across all channels
* `SendADC`: Sends the value from an ADC over a DMX channel
* `SendTestPackets`: Sends test packets (start code 55h)

Examples that show how to utilize synchronous and asynchronous transmission:
* `SIPSenderAsync`: Sends SIP packets asynchronously
* `SIPSenderSync`: Sends SIP packets synchronously

Examples that show how to use a synchronous packet handler in a receiver:
* `SIPHandler`: Understands System Information Packets (SIP) (start code CFh)
* `TextPacketHandler`: Understands text packets (start codes 17h and 90h)

Transmitter timing examples:
* `RegenerateDMX`: Regenerates received DMX onto a different serial port and
  with different timings

Other examples:
* `FastLEDController`: Demonstrates DMX pixel output using FastLED

A more complex example showing how to behave as a DMX USB Pro Widget is
in `USBProWidget`.

### Synchronous vs. asynchronous operation

Both transmission and reception operate asynchronously. This means that
there's potentially a continuous stream of data being sent or received in
the background.

The transmitter will keep sending the same data until it's changed externally
using one of the `Sender::set` functions. Similarly, `Receiver::readPacket` and
`Receiver::get` will return only the latest data; if more than a small amount of
time elapses between calls, then the data may have been changed more than once.

Both `Sender` and `Receiver` have a mode where they can operate synchronously.
With `Sender`, the stream can be paused and resumed, and packets inserted at
appropriate spots. Similarly, `Receiver` can send received packets as they
arrive to start code-specific instances of `Responder`.

## DMX receive

### Code example

First, create an object using one of the hardware serial ports, different from
any serial ports being used for transmission:

```c++
namespace teensydmx = ::qindesign::teensydmx;

teensydmx::Receiver dmxRx{Serial1};
```

Before using the instance, start the serial port internals:

```c++
dmxRx.begin();
```

Using your own buffer whose length is at least `len` bytes, check for a packet
having arrived, and if it has, copy `len` values starting from channel
`startChannel`:

```c++
// For this example, assume buf is at least len bytes long
int read = dmxRx.readPacket(buf, startChannel, len);
```

`read` will contain the number of bytes read, -1 if no packet is available, or
zero if no values were read.

Note that channel zero contains the start code, a special value, usually zero,
that occupies the first byte of the packet. The maximum DMX packet size is
513, but may be smaller, depending on the system.

Each call to `readPacket` is independent, meaning that if no packet has arrived
after a call to this function, subsequent calls will return -1.

### Retrieving 16-bit values

Retrieving 16-bit values is easy with one available 16-bit function:
1. `get16Bit` for single words.

This works the same as the 8-bit `get` function, but uses the `uint16_t`
type instead.

### Error counts and disconnection

The DMX receiver keeps track of three types of errors:

1. Packet timeouts.
2. Framing errors, including bad BREAKs.
3. Short packets, i.e. those packets that occupy less than 1196us.

The counts can be retrieved via `errorStats()`. This returns an `ErrorStats`
object containing each of the error counts. These metrics are reset to zero when
the receiver is started or restarted.

An associated concept is _disconnection_. A receiver is considered _connected_
when it is receiving valid DMX packets. When any timeouts occur, or when invalid
BREAKs are detected, then the receiver considers itself _disconnected_. As soon
as valid packets reappear, the receiver is once again _connected_.

To determine whether a receiver is connected, call its `connected()` function.
A callback function, to be notified when this event happens, can also be set
using `onConnectChange`.

The following example sets the built-in LED according to the current connected
state. It uses a lambda expression, but a normal function could be used instead.

```c++
dmxRx.onConnectChange([](Receiver *r) {
  digitalWriteFast(LED_BUILTIN, r.connected() ? HIGH : LOW);
});
```

#### The truth about connection detection

In actual fact, the connection detection only works for most of the cases where
timeouts occur or when bad BREAKs happen. The one thing it can't do is detect
when a BREAK condition occurs permanently, and, if none of the periodic
interrupt timers (PIT) are available, when an IDLE condition occurs permanently.
Unless more data comes in, no further UART interrupts will be generated, so it
is up to the user's code to detect this case.

The `Flasher` example contains an example of how to do this. In essence, the
`readPacket` function will return a positive value if the desired data is
available in the last packet received. A timer can be reset when valid data is
received, and then subsequent code can use the difference between the current
time and the timer value to determine if there's been a timeout.

An example that uses `elapsedMillis` to encapsulate the current time call:

```c++
constexpr uint32_t kTimeout = 1000;  // In milliseconds

elapsedMillis lastPacketTimer{0};
uint8_t buf[250];

void loop() {
  int read = dmxRx.readPacket(buf, 0, 217);
  if (read == 217) {  // Comparing to > 0 means that _some_ data was received
    // All the requested data was received
    // Do stuff with it
    lastPacketTimer = 0;
  }

  if (lastPacketTimer < kTimeout) {
    // Do work
  } else {
    // No connection
  }
}
```

A second technique would be to use the value returned from
`lastPacketTimestamp()` as something that closely approximates the true latest
timestamp. For example:

```c++
constexpr uint32_t kTimeout = 1000;  // In milliseconds

void loop() {
  if (millis() - dmxRx.lastPacketTimestamp() >= kTimeout) {
    // Respond to the timeout
  }
  // Do work
}
```

In actuality, a timer could be used to detect this condition, but it was chosen
to not do this because it would use up a timer, add one extra timeout-specifying
method, and the user code is probably good enough.

In summary, the _connected_ concept here has more to do with line noise and bad
timing than it does with a physical connection. Perhaps a future release will
rename this API concept or address it with the timer...

#### Keeping short packets

_Short packets_ are nothing more than valid-looking packets that span a duration
less than 1196us. The default is to discard them because they may indicate
line noise.

It is possible to keep the data from these packets by enabling the "Keep Short
Packets" feature using the `setKeepShortPackets` function. If these are kept
then the `PacketStats::isShort` variable will indicate whether the associated
packet is a _short packet_.

Recall that the `readPacket` function can atomically retrieve packet statistics
associated with the packet data. If `packetStats()` is used instead, then
there's no guarantee that the values will be associated with the most recent or
next packet data retrieval calls.

Note that there is also an `isKeepShortPackets` function that can be polled for
the current state of this feature.

### Packet statistics

Packet statistics are tracked and the latest can be retrieved from a
`PacketStats` object returned by `packetStats()`. Included are these variables:

1. `size`: The latest received packet size.
2. `isShort`: Whether the last packet was a _short packet_, a packet having a
   duration less than 1196us. Keeping these packets can be enabled with the
   `setKeepShortPackets` function.
3. `timestamp`: The timestamp of the last packet received, in milliseconds. This
   is set to the time the packet was recognized as a packet and not the end of
   the last stop bit.
4. `breakPlusMABTime`: The sum of the BREAK and MAB times, in microseconds. It's
   not possible to determine where the BREAK ends and the MAB starts without
   using another pin to watch the RX line. To set up a connected pin, see
   `setRXWatchPin` and the section above on
   [RX line monitoring](#receiver-timing-limitations-and-rx-line-monitoring).
5. `breakToBreakTime`: The latest measured BREAK-to-BREAK time, in microseconds.
   Note that this is not collected at the same time as the other variables and
   only represents the last known duration. This will be out of sync with the
   rest of the values in the presence of packet errors.
6. `frameTimestamp`: The BREAK start timestamp, in microseconds.
7. `packetTime`: The duration of the last packet, in microseconds, measured from
   BREAK start to the end of the last slot.
8. `breakTime`: The packet's BREAK time, set if RX line monitoring is enabled.
9. `mabTime`: The packet's MAB time, set if RX line monitoring is enabled.

If the RX line is not being monitored, then the BREAK and MAB times will be set
to zero.

There is also an optional parameter in `readPacket`, a `PacketStats*`, that
enables retrieval of this data atomically with the packet data.

These metrics are reset to zero when the receiver is started or restarted.

### Error statistics

Error statistics are tracked and the latest can be retrieved from an
`ErrorStats` object returned by `errorStats()`. Included are these counts:

1. `packetTimeoutCount`: Packet timeouts.
2. `framingErrorCount`: Framing error count, including BREAKs that were
   too short.
3. `shortPacketCount`: Packets that were too short.
4. `longPacketCount`: Packets that were too long.

### Synchronous operation by using custom responders

There is the ability to notify specific instances of `Responder` when packets
having specific start codes arrive. To implement the simplest form, simply
extend `Responder`, override the `receivePacket` function, and attach an
instance to one or more start codes using `Receiver::setResponder`.
`receivePacket` will be called for each packet received that has one of the
desired start codes.

As well, by default, handlers will "eat" packets so that they aren't available
to callers to the `Receiver` API. To change this behaviour, override
`Responder::eatPacket()`.

For example, let's say you want to change the local display when a text packet
arrvies. The following partial code example shows how to do this.

```c++
class TextHandler : public teensydmx::Responder {
 public:
  static constexpr uint8_t kStartCode = 0x17;

  void receivePacket(const uint8_t *buf, int len) override {
    // The packet must contain at least 3 bytes (plus the start code)
    if (len < 4) {
      return;
    }
    uint8_t page = buf[1];
    uint8_t charsPerLine = buf[2];

    // Some checks should be made here for the data not ending in a
    // NUL character

    // Assume the existence of this function
    setText(page, charsPerLine,
            reinterpret_cast<const char *>(&buf[3]), len - 3);
  }
}

TextHandler textHandler;

void setup() {
  // ...
  dmxRx.setResponder(TextHandler::kStartCode, &textHandler)
  // ...

  dmxRx.begin();
}
```

Responders can be added at any time.

Complete synchronous operation examples using SIP and text packets can be found
in `SIPHandler` and `TextPacketHandler`.

#### Responding

Protocols such as RDM need the ability, not only to process specific packets,
but to respond to them as well. Timing is important, so a `Responder`
implementation can also be notified of each byte as it arrives. The function of
interest is `processByte`.

As bytes are received, the implementation tracks some internal state. When it is
decided that a response is necessary, it returns a positive value indicating how
many bytes it placed into the output buffer, for transmitting back to the
transmitter. The `Responder` needs to implement `outputBufferSize()` in order
for any response to be sent. `processByte` will be passed a buffer at least as
large as the value returned from `outputBufferSize()`.

Some other functions that specify some timings should also be implemented.
Please consult the `Responder.h` documentation for more details.

Because all processing happens within an interrupt context, it should execute as
quickly as possible. Any long-running operations should be executed in the main
loop (or some other execution context). If the protocol allows for it, the
`Responder` can reply with a "not yet" response, and then return any queued
processing results when ready.

A more complete example is beyond the scope of this README.

## DMX transmit

### Code example

First, create an object using one of the hardware serial ports, different from
any serial ports being used for receive:

```c++
namespace teensydmx = ::qindesign::teensydmx;

teensydmx::Sender dmxTx{Serial2};
```

Before using the instance, optionally set up the packet size and refresh rate,
and then start the serial port internals to begin transmission:

```c++
// Optional: dmxTx.setPacketSize(100);
// Optional: dmxTx.setRefreshRate(40);
dmxTx.begin();
```

Set one or more channel values using one of the `set` functions:

```c++
// Set channel 6 to 219
dmxTx.set(6, 219);
```

The other `set` function can set multiple channels at once. This is left as an
exercise to the reader.

### Setting 16-bit values

Setting 16-bit values is easy with two available 16-bit functions:
1. `set16Bit` for single words, and
2. `set16Bit` for an array of words.

These work the same as the 8-bit `set` functions, but use the `uint16_t`
type instead.

### Packet size

The packet size can be adjusted and retrieved via `setPacketSize` and
`packetSize()`. Smaller packets will naturally result in a higher
refresh rate.

This can be changed at any time.

The minimum packet time is 1204us, and so as long as the actual time doesn't
fall short of this, there is no minimum packet size. The equation is:

```
Packet Size >= max{(1204us - BREAK - MAB)/44us, 0}
```

### Transmission rate

The transmission rate can be changed from a maximum of about 44Hz down to as low
as you wish. See the `setRefreshRate` and `refreshRate()` in `Sender`.

Note that the rate won't be higher than the limits dictated by the protocol,
about 44Hz, no matter how high it's set. The default is, in fact, `INFINITY`.

This can be changed at any time.

If the MBB time is also specified and it would conflict with the desired rate,
then the larger of the two possible MBB times will be used to achieve either the

CHANGELOG.md
# Changelog for the TeensyDMX Library

This document details the changes between each release.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added
* Provided an option to replace internal use of the Teensy library's
  `IntervalTimer` API with our own custom `PeriodicTimer` API (see
  v4.1.0-beta.2). This improves transmitted BREAK time accuracy (when using a
  timer and not serial parameters). Set the global `TEENSYDMX_USE_PERIODICTIMER`
  macro when building to use this timer API instead.
* Added a way to set the inter-slot MARK time. See `Sender::setInterSlotTime`
  and `Sender::interSlotTime()`.
* Added a way to set the MARK before BREAK (MBB) time. See `Sender::setMBBTime`
  and `Sender::mbbTime()`.
* New `RegenerateDMX` example.
* New `FastLEDController` example.
* Added `Sender::setPacketSizeAndData` for atomically setting the packet size
  and data. This doesn't grab the lock if the new packet size is the same.
* New `TeensyDMX::serialNumber()` function that returns the serial port number.

### Changed
* Changed relevant `__disable_irq()`/`__enable_irq()` pairs to
  `ATOMIC_BLOCK(ATOMIC_RESTORESTATE)` so that interrupts are not enabled
  inappropriately, for example if the calling program had disabled them.
* Clarified some details about minimum packet size.
* Improved MAB time measurement when using an RX watch pin by watching for the
  MAB fall time.
* Made `Sender` and `Receiver` movable.

### Fixed
* Allow 2% smaller character time when determining a bad break. This fixes a
  problem where packets with a slightly faster, but within allowed tolerance,
  baud rate were being discarded.
* Made `Responder::~Responder()` `virtual`.
* Fixed send handler code paths that shouldn't have called `setCompleting()`.
  This change was introduced in commit a2ac5f0.
* Fixed continuous TX interrupts in the transmitter (appears as a program
  freeze) when RX gets triggered.
  See: https://github.com/ssilverman/TeensyDMX/issues/20
* Fixed missing `disableRX()` calls in the send handlers.
* Changed receiver to only check for bad break at first byte. It wasn't
  technically accounting for inter-slot time, but that didn't actually matter;
  it still simplified the code.

## [4.2.0]

### Added
* New `Sender::fill` function that fills a channel range with a value.
* Added NULL checking for the `values` pointer in `Sender::set`
  and `Sender::set16Bit`.

### Changed
* No more volatile objects with weird copy constructors and assignment
  operators. Using atomic fences instead.
* Improved use of public, protected, private, and volatile.
* `Sender::setPacketSize` now disallows sizes of zero. The new allowed range
  is 1-513.
* Timeout detection in `Flasher`, `USBProWidget`, and the README now compare
  using strictly less than the timeout value so that the timeout value itself
  can be used to manually trigger a timeout, instead of having to use an
  arbitrary value greater than the timeout value.
* Changed uses of `std::shared_ptr` to regular pointers.

### Fixed
* Ensure the packet size can't be changed in the middle of sending a packet.
* Improved lambda hygiene.

## [4.1.1]

### Added
* Added a BREAK start timestamp to `PacketStats`, `frameTimestamp`. The value is
  in microseconds.
* Added `Receiver::rxWatchPin()` for retrieving the RX watch pin number.
* Added a note to the docs for `Receiver::setRXWatchPin` reminding the caller to
  configure the pin for input.

### Changed
* DMX TX and RX are now started with new `startRx` and `startTx` convenience
  functions in the `USBProWidget` example.
* Updated comment layout in the examples to be <= 70 columns to help improve
  readability, for example for larger fonts.

### Fixed
* Fixed `Chaser` example to start the blink and packet correctly.
* Fixed `TextPacketHandler` example to define one of the static members.
* Fixed `Sender` to use double-buffering to guarantee atomic behaviour.

## [4.1.0]

### Added
* `USBProWidget` example: Added received message error handling.
* Added 7- and 9-bit format support when using serial parameters to accomplish
  BREAK and MAB timing.
* `Sender::breakTime()` and `mabTime()` now return correct values if the sender
  is using serial parameters instead of a timer to generate the BREAK/MAB times.

### Changed
* `USBProWidget` example:
  * Moved DMX receive processing into the main loop.
  * Reduced the minimum returned BREAK time to 10.67us instead of 9*10.67us.
  * Updated uses of `memcpy` and `memcmp` to `std::copy_n` and `std::equal`,
    respectively.
* `BasicSend` example: Now calling `begin()` after setting the channel contents.
* `Sender::setBreakSerialParams` now checks for valid parameters and returns a
  Boolean indicating success.

### Fixed
* `USBProWidget` example:
  * More correct changed-DMX behaviour.
  * All commands except the appropriate ones now reset the program to
    input mode.
* Fixed `Sender::resumeFor` to ensure the UART has been started before setting
  it to active.

## [4.1.0-beta.2]

### Added
* Added a way to use either serial parameters or a timer to achieve the BREAK
  and MAB timings. Along with this, added the ability to change those
  serial parameters. New functions:
  * `setBreakSerialParams`
  * `breakSerialBaud`
  * `breakSerialFormat`
  * `setBreakUseTimerNotSerial`
  * `isBreakUseTimerNotSerial`
* Added an "Examples" section (under "How to use") to the README.
* Added a `USBProWidget` example.

### Changed
* Replaced use of the custom `PeriodicTimer` with the Teensy library's
  `IntervalTimer`. This helps compatibility with other libraries that use this
  API. This means that BREAK times are slightly less accurate (when using a
  timer and not serial parameters).
* Changed the default way to generate BREAK and MAB timings to use serial
  parameters instead of a timer.

src/main.cpp
// Basic main program, so it's easy to compile the project.
//
// This uses the concept of sub-sketches, where serial input selects which
// sub-sketch to run. The initial running sketch is the Null sketch; it does
// nothing. The other sketches are:
// 1. Chaser: This cycles an output of 255 on each channel.
// 2. Flasher: Receives input and flashes an LED with a period proportional to
//    the input.
//
// (c) 2018-2021 Shawn Silverman

// Define MAIN_TEST_PROGRAM to use this test program.
#ifdef MAIN_TEST_PROGRAM

// C++ includes
#include <cstdint>
#include <limits>

#include <Arduino.h>
#include "TeensyDMX.h"

namespace teensydmx = ::qindesign::teensydmx;

// Basic sub-sketch. The constructor and destructor don't do setup or teardown.
class Sketch {
 public:
  Sketch() = default;
  virtual ~Sketch() = default;

  virtual void setup() {}
  virtual void tearDown() {}
  virtual void loop() {}

  // Sketch info
  virtual String name() = 0;
  virtual int type() = 0;
};

// NullSketch supports the Null Pattern.
class NullSketch final : public Sketch {
 public:
  NullSketch() = default;
  ~NullSketch() override = default;

  String name() override {
    return "Null";
  }

  void setup() override;

  int type() override {
    return 'n';
  }
};

// Chaser cycles an output of 255 on each channel.
class Chaser final : public Sketch {
 public:
  Chaser();
  ~Chaser() override = default;

  String name() override {
    return "Chaser";
  }

  void setup() override;
  void tearDown() override;
  void loop() override;

  int type() override {
    return 'c';
  }

 private:
  static constexpr int kChannel = 1;
  static constexpr unsigned long kChaseInterval = 1000;  // 1s
  static constexpr int kPacketSize = 513;
  static constexpr float kRefreshRate = std::numeric_limits<float>::infinity();

  teensydmx::Sender dmx_;
  int channel_;
  elapsedMillis lastChaseTimer_;
};

// Flasher flashes the LED according to a DMX input speed.
class Flasher final : public Sketch {
 public:
  Flasher();
  ~Flasher() override = default;

  String name() override {
    return "Flasher";
  }

  void setup() override;
  void tearDown() override;
  void loop() override;

  int type() override {
    return 'f';
  }

 private:
  static constexpr int kChannel = 1;
  static constexpr unsigned long kTimeout = 1000;

  // Minimum and maximum frequencies, expressed as a period in milliseconds
  static constexpr int32_t kPeriodMax = 1000;
  static constexpr int32_t kPeriodMin = 30;

  teensydmx::Receiver dmx_;
  uint8_t lastValue_;
  elapsedMillis lastFrameTimer_;

  // Wave values
  int32_t period_;
  int64_t phi_;
};

// The LED pin.
constexpr uint8_t kLEDPin = LED_BUILTIN;

// Pin for enabling or disabling the transmitter.
constexpr uint8_t kTXPin = 17;

// The serial port to use.
HardwareSerial &uart = Serial3;

// The current sketch
NullSketch nullSketch{};
Chaser chaser{};
Flasher flasher{};
Sketch *currSketch = &nullSketch;

// ---------------------------------------------------------------------------
//  Main program
// ---------------------------------------------------------------------------

// Changes the current sketch, if different. This ignores any unknown
// sketch type.
void changeSketch(int sketchType);

// Main program setup.
void setup() {
  // Initialize the serial port
  Serial.begin(115200);
  while (!Serial && millis() < 4000) {
    // Wait for initialization to complete or a time limit
  }
  Serial.println("Starting main program.");

  // Set up any pins
  pinMode(kLEDPin, OUTPUT);
  digitalWriteFast(kLEDPin, HIGH);  // Start with the LED on
  pinMode(kTXPin, OUTPUT);
  digitalWriteFast(kTXPin, LOW);

  // Initialize with the Null sketch
  currSketch = &nullSketch;
  currSketch->setup();

  Serial.println("Hello, DMX World!");
}

// Prints a sketch and prepends a star if it's the current active one.
// This indents by four spaces.
void printSketch(int sketchType, const char *s) {
  if (currSketch->type() == sketchType) {
    Serial.print("   *");
  } else {
    Serial.print("    ");
  }
  Serial.println(s);
}

// Main program loop.
void loop() {
  // Potentially choose a new sketch
  int avail = Serial.available();
  if (avail > 0) {
    int b = Serial.read();

    // First, transform aliases
    // and perform any commands
    switch (b) {
      case 'C':
        b = 'c';
        break;
      case 'F':
        b = 'f';
        break;
      case 'N':
        b = 'n';
        break;
      case '?':
        b = -1;
        Serial.println("Sketches:");
        printSketch('c', "(C)haser");
        printSketch('f', "(F)lasher");
        printSketch('n', "(N)ull");
        break;
    }

    if (b >= 0) {
      changeSketch(b);
    }
  }

  currSketch->loop();
}

// ---------------------------------------------------------------------------
//  Support functions
// ---------------------------------------------------------------------------

extern "C" void *_sbrk(int incr);

// Returns the amount of free RAM.
// See: https://forum.pjrc.com/threads/57269-I-can-t-seem-to-use-all-of-the-Teensy-4-0-s-memory
// See: https://github.com/arduino/ArduinoCore-API/issues/73
int freeRAM() {
#if defined(__IMXRT1052__) || defined(__IMXRT1062__)
  extern unsigned long _heap_end;
  intptr_t end = reinterpret_cast<intptr_t>(&_heap_end);
#else
  int v;
  intptr_t end = reinterpret_cast<intptr_t>(&v);
#endif  // __IMXRT1062__ || __IMXRT1052__
  return end - reinterpret_cast<intptr_t>(_sbrk(0));
}

void changeSketch(int sketchType) {
  if (sketchType == currSketch->type()) {
    return;
  }

  // Hold onto the old sketch so we can destroy it
  Sketch *oldSketch = currSketch;

  // Create a new sketch
  switch (sketchType) {
    case 'c':
      currSketch = &chaser;
      break;
    case 'f':
      currSketch = &flasher;
      break;
    case 'n':
      currSketch = &nullSketch;
      break;
    default:
      return;
  }

  Serial.printf("Changing sketch to: %s\n", currSketch->name().c_str());
  oldSketch->tearDown();
  currSketch->setup();
  Serial.printf("Free RAM: %d\n", freeRAM());
}

// ---------------------------------------------------------------------------
//  Sub-sketches
// ---------------------------------------------------------------------------

void NullSketch::setup() {
  digitalWriteFast(kLEDPin, HIGH);
}

Chaser::Chaser()
    : dmx_{uart},
      channel_(kChannel),
      lastChaseTimer_(0) {}

void Chaser::setup() {
  digitalWriteFast(kLEDPin, LOW);
  digitalWriteFast(kTXPin, HIGH);

  dmx_.set(1, 255);
  for (int i = 2; i < 513; i++) {
    dmx_.set(i, 0);
  }
  dmx_.setPacketSize(kPacketSize);
  dmx_.setRefreshRate(kRefreshRate);
  dmx_.begin();

  channel_ = 1;
  lastChaseTimer_ = 0;
}

void Chaser::tearDown() {
  digitalWriteFast(kTXPin, LOW);

  dmx_.end();
}

void Chaser::loop() {
  if (lastChaseTimer_ < kChaseInterval) {
    return;
  }
  lastChaseTimer_ = 0;

  // Set the current channel to zero, advance it, and then set the new channel
  // to 255
  dmx_.set(channel_, 0);
  channel_++;
  if (channel_ >= dmx_.packetSize()) {
    channel_ = 1;
  }
  dmx_.set(channel_, 255);

  // Blink the LED
  digitalWriteFast(kLEDPin, HIGH);
  delay(20);
  digitalWriteFast(kLEDPin, LOW);
}

Flasher::Flasher()
    : dmx_{uart},
      lastValue_(0),
      lastFrameTimer_(0),
      period_(kPeriodMax),
      phi_(0) {}

void Flasher::setup() {
  digitalWriteFast(kLEDPin, HIGH);  // Start with the LED on
  digitalWriteFast(kTXPin, LOW);

  dmx_.begin();
  lastFrameTimer_ = kTimeout;
}

void Flasher::tearDown() {
  digitalWriteFast(kLEDPin, LOW);

  dmx_.end();
}

void Flasher::loop() {
  uint8_t value;
  int read = dmx_.readPacket(&value, kChannel, 1);
  int64_t t = millis();
  if (read > 0) {
    // We've read everything we want to
    lastValue_ = value;
    lastFrameTimer_ = 0;

    // Use a wave equation to make the speed-ups and slow-downs smoother using
    // the offset, phi
    int32_t newPeriod = map(lastValue_, 0, 255, kPeriodMax, kPeriodMin);
    if (newPeriod != period_) {
      phi_ = (t*(period_ - newPeriod) + newPeriod*phi_)/period_;
      period_ = newPeriod;
    }
  }

  if (lastFrameTimer_ < kTimeout) {
    int32_t v = (t - phi_)%period_;
    if (v < period_/2) {
      digitalWriteFast(kLEDPin, HIGH);
    } else {
      digitalWriteFast(kLEDPin, LOW);
    }
  } else {
    digitalWriteFast(kLEDPin, LOW);
  }
}

#endif  // MAIN_TEST_PROGRAM
